scripts/remoteperformance.ps1

```diff
@@ -0,0 +1,298 @@
+            }
+        }
+    } catch {
+        $RetObj.apipaInterfaces = $null
+    }
+
+    return $RetObj
+}
+
+function LocalTeardown {
+    param ($LocalCache)
+    if ($null -ne $LocalCache.apipaInterfaces) {
+        # Re-enable the interfaces we disabled earlier.
+        Write-Debug "Re-enabling APIPA interfaces"
+        $LocalCache.apipaInterfaces | Enable-NetAdapter
+    }
+}
+
+function Invoke-Test {
+    param ($Test)
+
+    Write-Output "Running Test $Test"
+
+    $RemoteExe = Get-ExeName -PathRoot $RemoteDirectory -Platform $RemotePlatform -IsRemote $true -TestPlat $Test.Remote
+    $LocalExe = Get-ExeName -PathRoot $LocalDirectory -Platform $LocalPlatform -IsRemote $false -TestPlat $Test.Local
+
+    # Check both Exes
+    $RemoteExeExists = Invoke-TestCommand -Session $Session -ScriptBlock {
+        param ($RemoteExe)
+        Test-Path $RemoteExe
+    } -ArgumentList $RemoteExe
+
+    $LocalExeExists = Test-Path $LocalExe
+
+    if (!$RemoteExeExists -or !$LocalExeExists) {
+        Write-Output "Failed to Run $Test because of missing exe"
+        if (!$RemoteExeExists) {
+            Write-Output "Missing Remote Exe $RemoteExe"
+        }
+        if (!$LocalExeExists) {
+            Write-Output "Missing Local Exe $LocalExe"
+        }
+        return
+    }
+
+    $LocalArguments = $Test.Local.Arguments.GetArguments().Replace('$RemoteAddress', $RemoteAddress)
+    $LocalArguments = $LocalArguments.Replace('$LocalAddress', $LocalAddress)
+
+    $CertThumbprint = Invoke-TestCommand -Session $Session -ScriptBlock {
+        return $env:QUICCERT
+    }
+
+    $RemoteArguments = $Test.Remote.Arguments.GetArguments().Replace('$Thumbprint', $CertThumbprint)
+
+    $RemoteJob = Invoke-RemoteExe -Exe $RemoteExe -RunArgs $RemoteArguments
+
+    $ReadyToStart = Wait-ForRemoteReady -Job $RemoteJob
+
+    if (!$ReadyToStart) {
+        Write-Output "Test Remote for $Test failed to start"
+        Stop-Job -Job $RemoteJob
+        return
+    }
+
+    $AllRunsResults = @()
+
+    try {
+        1..$Test.Iterations | ForEach-Object {
+            $LocalResults = Invoke-LocalExe -Exe $LocalExe -RunArgs $LocalArguments
+
+            $LocalParsedResults = Get-TestResult -Results $LocalResults -Matcher $Test.ResultsMatcher
+
+
+            $AllRunsResults += $LocalParsedResults
+
+            Write-Output "Run $($_): $LocalParsedResults kbps"
+            $LocalResults | Write-Debug
+        }
+    } finally {
+        $RemoteResults = Wait-ForRemote -Job $RemoteJob
+        Write-Debug $RemoteResults.ToString()
+    }
+
+    if ($Record) {
+        Get-RemoteFile -From ($RemoteExe + ".remote.etw") -To (Join-Path $OutputDir $Test.ToString() + ".etw")
+    }
+
+    Publish-TestResults -Test $Test `
+                        -AllRunsResults $AllRunsResults `
+                        -CurrentCommitHash $CurrentCommitHash `
+                        -OutputDir $OutputDir
```

Comment:
Only if `$Publish` is `$true`?

Reply:
No, we still want to do parts of this, as this also does the printing. Internally it checks for Publish before actually writing the file.

Reply:
So it's not just publishing results? We should rename this to a more appropriate name then.

Reply:
So, for PS functions to not complain when they get imported as a module, you have to use an approved list of verbs. 

https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7

I can't really find a better one in this list. And it does 2 things. It publishes to the output, and it optionally publishes the results file. So in a way, it is publish anyway.

Reply:
Ok.